__tests__/browser.test.js

```diff
@@ -0,0 +1,187 @@
+/* global fetchHar, page, SERVER_URL */
+const harExamples = require('har-examples');
+const jsonWithAuthHar = require('./__fixtures__/json-with-auth.har.json');
+const urlEncodedWithAuthHar = require('./__fixtures__/urlencoded-with-auth.har.json');
+
+describe('#constructRequest', () => {
+  beforeAll(async () => {
+    await page.exposeFunction('har_cookies', () => harExamples.cookies);
+    await page.exposeFunction('har_full', () => harExamples.full);
+    await page.exposeFunction('har_jsonWithAuthHar', () => jsonWithAuthHar);
+    await page.exposeFunction('har_multipartData', () => harExamples['multipart-data']);
+    await page.exposeFunction('har_multipartFile', () => harExamples['multipart-file']);
+    await page.exposeFunction('har_multipartFormData', () => harExamples['multipart-form-data']);
+    await page.exposeFunction('har_textPlain', () => harExamples['text-plain']);
+    await page.exposeFunction('har_urlEncodedWithAuthHar', () => urlEncodedWithAuthHar);
+  });
+
+  beforeEach(async () => {
+    await page.goto(SERVER_URL, { waitUntil: 'load' });
+
+    // Since `constructRequest` returns a `Request` object, which is not serializable, for our tests we need to have
+    // Puppeteer return an object containing the properties and data we're looking for in order for us to be able to run
+    // assertions within Jest.
+    //
+    // Why not use `page.exposeFunction`? That method unfortunately serializes all parameters that are supplied to
+    // callbacks created with it, and as `Request` can't be serialized we'll end up with an empty object. Also since
+    // we don't need to pass data from the Node layer into Puppetter, as you normally would with `exposeFunction`, we
+    // can skirt that and simply create what we need on the page instance.
+    //
+    // https://github.com/puppeteer/puppeteer/issues/1750
+    // https://pptr.dev/#?product=Puppeteer&version=v5.2.1&show=api-pageevaluatepagefunction-args
+    await page.evaluate(() => {
+      window.parseRequest = async req => {
+        const headers = {};
+        // eslint-disable-next-line no-restricted-syntax
+        for (const header of req.headers.entries()) {
+          headers[header[0]] = header[1];
```

Comment:
There might be a good use case for `fromEntries` [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries)

Reply:
Great idea.